libcnb/src/tracing.rs

```diff
@@ -0,0 +1,136 @@
+        .build();
+
+    global::set_tracer_provider(provider.clone());
+
+    let tracer = provider.versioned_tracer(lib_name, lib_version, None as Option<&str>, None);
+
+    let mut span = tracer.start(trace_name);
+    span.set_attributes(vec![
+        KeyValue::new("buildpack_id", buildpack.id.to_string().clone()),
+        KeyValue::new("buildpack_name", buildpack.name.clone().unwrap_or_default()),
+        KeyValue::new("buildpack_version", buildpack.version.to_string()),
+        KeyValue::new(
+            "buildpack_homepage",
+            buildpack.homepage.clone().unwrap_or_default(),
+        ),
+    ]);
+    BuildpackTrace { provider, span }
+}
+
+impl BuildpackTrace {
+    pub(crate) fn set_error(&mut self, err: &dyn std::error::Error) {
+        self.span.set_status(Status::error(err.to_string()));
+        self.span.record_error(err);
+    }
+    pub(crate) fn add_event(&mut self, name: &'static str) {
+        self.span.add_event(name, vec![]);
+    }
+}
+
+impl Drop for BuildpackTrace {
+    fn drop(&mut self) {
+        self.span.end();
+        self.provider.force_flush();
+        global::shutdown_tracer_provider();
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::start_trace;
+    use libcnb_data::buildpack::{Buildpack, BuildpackVersion};
+    use std::{
+        collections::HashSet,
+        fs,
+        io::{Error, ErrorKind},
+    };
+
+    #[test]
+    fn test_tracing() {
+        let buildpack = Buildpack {
+            id: "company.com/foo"
+                .parse()
+                .expect("Valid BuildpackId should parse"),
+            version: BuildpackVersion::new(0, 0, 0),
+            name: Some("Foo buildpack for company.com".to_string()),
+            homepage: None,
+            clear_env: false,
+            description: None,
+            keywords: vec![],
+            licenses: vec![],
+            sbom_formats: HashSet::new(),
+        };
+        let phase = "bar";
+        let event = "baz-event";
+        let error_message = "its broken";
+        let telemetry_path = "/tmp/cnb-telemetry/company_com_foo-bar.jsonl";
+        _ = fs::remove_file(telemetry_path);
+
+        {
+            let mut trace = start_trace(&buildpack, phase);
+            trace.add_event(event);
+            trace.set_error(&Error::new(ErrorKind::Other, error_message));
+        }
+        let tracing_contents = fs::read_to_string(telemetry_path)
+            .expect("Expected telemetry file to exist, but couldn't read it");
+
+        println!("tracing_contents: {tracing_contents}");
+        assert!(tracing_contents.contains(phase));
+        assert!(tracing_contents.contains(event));
+        assert!(tracing_contents.contains(error_message));
+        assert!(tracing_contents.contains(buildpack.id.as_str()));
+        assert!(tracing_contents.contains(&buildpack.version.to_string()));
+        assert!(
+            tracing_contents.contains(&buildpack.name.expect("Expected buildpack.name to exist"))
+        );
```

Comment:
I'm a bit uneasy about this test. It doesn't actually test that the fields are set correctly, or that the file is valid JSON - only that substrings exist in the file in general.

Reply:
IMO, the keys/fields are the business of `opentelemetry`, and I don't think we should be re-testing it's serialization. 
I don't mind validating that it is JSON, though. Additonally, traversing to the values is a bit laborious, since object serialization looks like `[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]` rather than something a little more JSON-native like `{resource-attr: "resource-attr-val-1"}`.

Reply:
Yeah, fair enough on those points. I can't check the entire full file contents because 1) the keys are not consistently ordered and 2) the payload has timestamps for spans and events. I did add some rigor to these tests in 74a542092172fa26eccb8f5a6554e45707456850, though. It's now a bit more rigorous, checking text against expected keys and values. I feel pretty good about this balance now, especially given the test also validates the json.

Reply:
The new assertions look great, thank you!